shell/server/core.js

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2014 Sandstorm Development Group, Inc. and contributors
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+var Crypto = Npm.require("crypto");
+var Capnp = Npm.require("capnp");
+
+var HandlePersistent = Capnp.importSystem("sandstorm/supervisor.capnp").HandlePersistent;
+var SandstormCore = Capnp.importSystem("sandstorm/supervisor.capnp").SandstormCore;
+var SandstormCoreFactory = Capnp.importSystem("sandstorm/backend.capnp").SandstormCoreFactory;
+var OngoingNotificationPersistent = Capnp.importSystem("sandstorm/supervisor.capnp").OngoingNotificationPersistent;
+
+function SandstormCoreImpl(grainId) {
+  this.grainId = grainId;
+}
+
+var makeSandstormCore = function (grainId) {
+  return new Capnp.Capability(new SandstormCoreImpl(grainId), SandstormCore);
+};
+
+function NotificationHandle(notificationId) {
+  this.notificationId = notificationId;
+}
+
+function makeNotificationHandle(notificationId) {
+  return new Capnp.Capability(new NotificationHandle(notificationId), HandlePersistent);
+}
+
+function dropWakelock(grainId, wakeLockNotificationId) {
+  // For some reason, Mongo returns an object that looks buffer-like, but isn't a buffer.
+  // We have to explicitly copy it before we can pass it on to node-capnp.
+  var copiedId = new Buffer(wakeLockNotificationId);
+  waitPromise(openGrain(grainId).supervisor.drop({wakeLockNotification: copiedId}).catch(function (err) {
+    if (shouldRestartGrain(err, 0)) {
+      return openGrain(grainId, true).supervisor.drop({wakeLockNotification: copiedId});
+    }
+  }));
+}
+
+function dismissNotification(notificationId) {
+  var notification = Notifications.findOne({_id: notificationId});
+  if (notification) {
+    Notifications.remove({_id: notificationId});
+    if (notification.ongoing) {
+      // For some reason, Mongo returns an object that looks buffer-like, but isn't a buffer.
+      // Only way to fix seems to be to copy it.
+      dropToken(notification.grainId, new Buffer(notification.ongoing));
+    }
+  }
+}
+
+function tryDropNotification(notificationId) {
+  // TODO(soon): This has serious timing issues with Mongo, since we're relying on it for refcounting
+  // Typically a notification handle will be saved and then immediately closed by the caller.
+  // This means that if Mongo doesn't actually find the ApiToken that was just stored, this method
+  // will dismiss the notification even though it was meant to be persisted.
```

Comment:
I think this timing issue can be solved by having `NotificationHandle` contain a boolean field indicating whether `save()` was called. Then, in `close()`, check for that boolean rather than searching the `ApiTokens` table.

It looks like the other place `tryDropNotification()` is called (inside `dropToken()`), there's no serious timing issue.

Or am I misunderstanding?